File: 多线程/printInOrder.py
from threading import Lock


class Foo:
    def __init__(self):
        self.firstEnd = Lock()
        self.secondEnd = Lock()
        self.firstEnd.acquire()  # 获取锁
        self.secondEnd.acquire()  # 获取锁

    def first(self, printFirst: 'Callable[[], None]') -> None:
        # printFirst() outputs "first". Do not change or remove this line.
        printFirst()
        self.firstEnd.release()  # 释放锁

    def second(self, printSecond: 'Callable[[], None]') -> None:
        # printSecond() outputs "second". Do not change or remove this line.
        with self.firstEnd:  # 等待锁被释放 或者 写 self.firstEnd.acquire()  # 获取锁 也可以
            printSecond()
            self.secondEnd.release()

    def third(self, printThird: 'Callable[[], None]') -> None:
        # printThird() outputs "third". Do not change or remove this line.
        with self.secondEnd:  # 等待锁被释放 或者 写 self.secondEnd.acquire()  # 获取锁 也可以
            printThird()

File: 多线程/printInOrder.c
// This header is part of the thread support library.
#include <semaphore.h>

class Foo {

protected:
    sem_t firstJobDone;
    sem_t secondJobDone;

public:

    Foo() {
        // 初始化
        // int sem_init(sem_t *sem,int pshared,unsigned int value);
        // pshared，指明信号共享情况（0，进程内的线程共享；非0，进程之间共享，并应该定位共享内存区域）
        // value，设置初始信号量
        // sem_init() 成功时返回 0；错误时，返回 -1，并把 errno 设置为合适的值。
        sem_init(&firstJobDone, 0, 0);
        sem_init(&secondJobDone, 0, 0);
    }

    void first(function<void()> printFirst) {
        // printFirst() outputs "first".
        printFirst();
        sem_post(&firstJobDone);  // firstJobDone 资源增加1
    }

    void second(function<void()> printSecond) {
        sem_wait(&firstJobDone);  // firstJobDone 资源减少1
        // printSecond() outputs "second".
        printSecond();
        sem_post(&secondJobDone);  // secondJobDone 资源增加1

    }

    void third(function<void()> printThird) {
        sem_wait(&secondJobDone); // secondJobDone 资源减少1
        // printThird() outputs "third".
        printThird();
    }
};
